app/src/main/java/com/keylesspalace/tusky/components/domainblocks/DomainBlocksFragment.kt

```diff
@@ -0,0 +1,108 @@
+import com.keylesspalace.tusky.di.Injectable
+import com.keylesspalace.tusky.di.ViewModelFactory
+import com.keylesspalace.tusky.util.hide
+import com.keylesspalace.tusky.util.show
+import com.keylesspalace.tusky.util.viewBinding
+import com.keylesspalace.tusky.util.visible
+import kotlinx.coroutines.flow.collectLatest
+import kotlinx.coroutines.launch
+import javax.inject.Inject
+
+class DomainBlocksFragment : Fragment(R.layout.fragment_domain_blocks), Injectable {
+
+    @Inject
+    lateinit var viewModelFactory: ViewModelFactory
+
+    private val binding by viewBinding(FragmentDomainBlocksBinding::bind)
+
+    private val viewModel: DomainBlocksViewModel by viewModels { viewModelFactory }
+
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        val adapter = DomainBlocksAdapter(viewModel::unblock)
+
+        binding.recyclerView.setHasFixedSize(true)
+        binding.recyclerView.addItemDecoration(DividerItemDecoration(view.context, DividerItemDecoration.VERTICAL))
+        binding.recyclerView.adapter = adapter
+        binding.recyclerView.layoutManager = LinearLayoutManager(view.context)
+
+        viewLifecycleOwner.lifecycleScope.launch {
+            viewModel.uiEvents.collect { event ->
+                when (event) {
+                    is DomainBlockEvent.UnblockError -> showUnmuteError(event.domain)
+                    is DomainBlockEvent.BlockError -> showMuteError(event.domain)
+                    is DomainBlockEvent.BlockSuccess -> showUnmuteSuccess(event.domain)
+                }
+            }
+        }
+
+        lifecycleScope.launch {
+            viewModel.pager.collectLatest { pagingData ->
+                adapter.submitData(pagingData)
+            }
+        }
+
+        adapter.addLoadStateListener { loadState ->
+            binding.progressBar.visible(loadState.refresh == LoadState.Loading && adapter.itemCount == 0)
+
+            if (loadState.refresh is LoadState.Error) {
+                binding.recyclerView.hide()
+                binding.messageView.show()
+                val errorState = loadState.refresh as LoadState.Error
+                binding.messageView.setup(errorState.error) { adapter.retry() }
+                Log.w(TAG, "error loading blocked domains", errorState.error)
+            } else if (loadState.refresh is LoadState.NotLoading && adapter.itemCount == 0) {
+                binding.recyclerView.hide()
+                binding.messageView.show()
+                binding.messageView.setup(R.drawable.elephant_friend_empty, R.string.message_empty)
+            } else {
+                binding.recyclerView.show()
+                binding.messageView.hide()
+            }
+        }
+    }
+
+    private fun showUnmuteError(domain: String) {
+        showSnackbar(
+            getString(R.string.error_unblocking_domain, domain),
+            R.string.action_retry
+        ) { viewModel.unblock(domain) }
+    }
```

Comment:
In the Notifications code the error classes have `@StringRes val message: Int` properties so that the code using the error doesn't need to worry about it.

They also contain the action that failed, so the retry code can be generic.

They also contain the throwable that caused the error, so more accurate information can be shown to the user.

Any reason not to do that here too?